### PR TITLE
Collective selection infrastructure for blocking collectives

### PR DESCRIPTION
**Pull request by AlexeyMalkhanov**
Features:
1. Collective selection infrastructure has been spread across all blocking collectives
2. New collective naming scheme: ```<MPI_Coll>__<intra|inter>__<algorithm>```
3. Code cleanup

### CODE REVIEW

**Comment by AlexeyMalkhanov** [comment]
@raffenet could you please launch the tests and review the content

File: src/mpi/coll/allgather/allgather.c [hunk 1]
```diff
@@ -101,10 +101,10 @@ int MPI_Allgather(const void *sendbuf, int sendcount, MPI_Datatype sendtype, voi
 #define MPI_Allgather PMPI_Allgather
 
 #undef FUNCNAME
-#define FUNCNAME MPIR_Allgather_intra_auto
+#define FUNCNAME MPIR_Allgather__intra__auto
```
Comment (raffenet) [inline src/mpi/coll/allgather/allgather.c]: @AlexeyMalkhanov what is the purpose of these changes to double underscores? If this PR is to be reviewed, there needs to be separation of functional and non-functional changes into separate commits. Commit messages need to be added to explain what is going on and why.

Comment (raffenet) [reply src/mpi/coll/allgather/allgather.c]: To be clear, it's not just functional vs. non-functional stuff. Commits should be independent and explained in the commit message. It's necessary for a decent code review.

**Comment by AlexeyMalkhanov** [comment]
@raffenet I've split the PR into two parts. The first stands for unified collectives naming scheme implementation where double underscores are separators between collective name, communicator type (where this collective will be applied) and algorithm name. The second commit extends new collective selection infrastructure to the rest of blocking collectives.

**Comment by AlexeyMalkhanov** [comment]
@raffenet I've updated the PR in accordance with the latest changes caused by code-cleanup.sh utility application

**Comment by AlexeyMalkhanov** [comment]
@raffenet @wesbland It turned out that whitespace-checker failed on some Makefile.mk files. It reports that we should use spaces instead of tabs. Since the checker is applied only to new lines it ignored the rest of lines which start with a tab, not with space. Could you please advise what we should do: ignore whitespace-checker results for makefiles or update every makefile substituting tabs to spaces?

**Comment by wesbland** [comment]
The whitespace checker (code formatting really) shouldn't be checking Makefiles. It should only be checking files with `.c, .h, .c.in, .h.in`.

**Comment by AlexeyMalkhanov** [comment]
@wesbland "git diff pmodels/master --check" shows only Makefile.mk files. How can I get the list of .c, .h, .c.in, .h.in files to be corrected?

**Comment by AlexeyMalkhanov** [comment]
@raffenet can we ignore the whitespace-checker errors in Makefile.mk files? If yes, could you please launch the tests for the PR?

**Comment by AlexeyMalkhanov** [comment]
@wesbland @raffenet Is it possible to review this PR for mpich-3.3b2? Despite the huge amount of the code in the PR it stands for spreading the same selection logic for all blocking collectives and unification of names for algorithms related objects like ids, containers, algorithm names, etc.

File: src/mpi/coll/allgather/allgather.c (hunk 1 above)
Comment (pavanbalaji) [reply src/mpi/coll/allgather/allgather.c]: @ddurnov I don't think double underscores are good for function names.  The functions that are missing intra/inter are algorithms which work in both cases.  Perhaps we could use `allcomm` to suggest both intra/inter instead?

**Comment by pavanbalaji** [comment]
At least the first patch (naming) can easily go into `3.3b2`.  I can't speak for the second patch right now.

File: src/mpi/coll/allgather/allgather.c (hunk 1 above)
Comment (ddurnov) [reply src/mpi/coll/allgather/allgather.c]: @pavanbalaji So should we have something like: ```<MPI_Coll>_<intracomm|intercomm|allcomm>_<algorithm name>``` in that case? In the end we should have a collective operation registration infrastructure so name won't matter that much (CVAR like idea) but right now formal fixed naming scheme would be helpful. (We may fetch all available collectives from the code by using simple pattern)

Comment (pavanbalaji) [reply src/mpi/coll/allgather/allgather.c]: I was recommending `<MPI_Coll>_<intra|inter|allcomm>_<algorithm_name>`.

Comment (ddurnov) [reply src/mpi/coll/allgather/allgather.c]: Thanks. There was a bit of confusion on our side because of ```__``` for container structure names and collective operation ids you have added.
Would be great if we could align container names/ids with collective operation names.

Comment (pavanbalaji) [reply src/mpi/coll/allgather/allgather.c]: OK, if that's what people prefer, that's fine with me.  But please get some more votes on the naming before the change.

Comment (AlexeyMalkhanov) [reply src/mpi/coll/allgather/allgather.c]: @raffenet @wesbland @alexander-sannikov @akhillanger Do you prefer ```__``` or ```_``` in the names of containers, ids and functions?

Comment (wesbland) [reply src/mpi/coll/allgather/allgather.c]: It doesn't make much difference to me now that the double underscore is already being used. In general, I think having the double underscore is a great way to get typos in your function names, but I guess that's what a good linter is for.

Comment (ddurnov) [reply src/mpi/coll/allgather/allgather.c]: Seems like ```_``` (single underscore) is the way

**Comment by AlexeyMalkhanov** [comment]
@pavanbalaji @raffenet @wesbland In accordance with the conversation about the naming scheme we have refactored the PR. Could you please take a look?

**Comment by wesbland** [comment]
The first patch committed a dirty UCX submodule. That needs to get reverted.

**Comment by wesbland** [comment]
The third patch (`<Gatherv|Scatterv>_allcomm_linear -> <Gatherv|Scatterv>_<inter|intra>_linear separation`) is pretty confusing because it mixes splitting one file into two and doing code reformatting. That makes it hard to tell which part is just refactoring and which part is code cleanup.

I'm not sure any of the code cleanup is actually necessary, but can it at least be a separate patch so we can review it separately?

Regardless of all that, should we be removing the original file or duplicating it (and perhaps moving common code into other functions? Aren't the new files new algorithms?

**Comment by wesbland** [comment]
(Sorry these comments are all showing up outside a review, but there's not a good way to talk about the patches as a whole in the GitHub review system)

All of the commit messages need to be shortened to fit in 50 characters (https://chris.beams.io/posts/git-commit/).

File: src/mpid/ch4/netmod/ofi/ofi_coll.h [hunk 2]
```diff
@@ -23,16 +23,27 @@
...
-
     nm_algo_parameters_container_out =
         MPIDI_OFI_Barrier_select(comm_ptr, errflag, (MPIDI_OFI_coll_algo_container_t *)
                                  ch4_algo_parameters_container_in);
-    mpi_errno = MPIDI_OFI_Barrier_call(comm_ptr, errflag, nm_algo_parameters_container_out);
+
+    switch (nm_algo_parameters_container_out->id) {
+        case MPIDI_OFI_Barrier_intra_dissemination_id:
+            mpi_errno =
+                MPIDI_OFI_Barrier_intra_dissemination(comm_ptr, errflag,
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_coll.h]: I'm confused about why we're duplicating all of the `MPIR` algorithms here. Can't we just call them directly instead of using these indirection functions?

File: src/mpid/ch4/netmod/ofi/ofi_coll_globals_default.c [hunk 3]
```diff
@@ -25,11 +25,136 @@ const MPIDI_OFI_coll_algo_container_t OFI_Reduce_intra_binomial_cnt = {
     .id = MPIDI_OFI_Reduce_intra_binomial_id
 };
 
-/* Allreduce default OFI containers initialization*/
+/* Allreduce default OFI containers initialization */
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_coll_globals_default.c]: Whitespace only changes should either **be removed** or done separately.

File: src/mpid/ch4/netmod/ofi/ofi_coll_params.h [hunk 4]
```diff
@@ -6,7 +6,7 @@ typedef enum {
 } MPIDI_OFI_Barrier_id_t;
 
 typedef union {
-    /* reserved for parameters related to SHM specific collectives */
+    /* reserved for parameters related to NETMOD specific collectives */
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_coll_params.h]: Typo changes should be a separate patch from substantive changes.

File: src/mpid/ch4/netmod/ofi/ofi_coll_params.h [hunk 5]
```diff
@@ -15,11 +15,11 @@ typedef union {
 typedef enum {
     MPIDI_OFI_Bcast_intra_binomial_id,
     MPIDI_OFI_Bcast_intra_scatter_recursive_doubling_allgather_id,
-    MPIDI_OFI_Bcast_intra_scatter_ring_allgather_id
+    MPIDI_OFI_Bcast_intra_scatter_ring_allgather_id,
```
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_coll_params.h]: This is a code formatting only change and should be a separate patch (or dropped).

File: src/mpid/ch4/netmod/ofi/ofi_coll.h (hunk 2 above)
Comment (AlexeyMalkhanov) [reply src/mpid/ch4/netmod/ofi/ofi_coll.h]: @wesbland @ddurnov  Till we don't have ofi specific implementations these functions are wrappers over MPIR algorithms. The current approach represents the skeleton of layered scheme for collective selection.

Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_coll.h]: I understand that there will be more in the future, but it seems list a ton of unnecessary code just to wrap the `MPIR` functions.

**Comment by pavanbalaji** [comment]
@AlexeyMalkhanov I looked at the first patch, and it's not consistent.  It's doing more than adding `allcomm` to the names, e.g., changing `recursive_doubling` to `dissemination`.  Can that please be separated out?  I haven't looked at the other patches yet.

File: src/mpid/ch4/src/ch4_coll_globals_default.c.in [hunk 6]
```diff
@@ -24,6 +24,12 @@ const MPIDI_coll_algo_container_t CH4_Barrier_intra_composition_beta_cnt = {
                                                    .barrier = -1}
 };
 
+const MPIDI_coll_algo_container_t CH4_Barrier_intra_composition_gamma_cnt = {
+    .id = MPIDI_CH4_Barrier_intra_composition_gamma_id,
+    .params.ch4_barrier_params.ch4_barrier_gamma = {
+                                                    .barrier = -1}
```
Comment (ddurnov) [inline src/mpid/ch4/src/ch4_coll_globals_default.c.in]: Would be great if we could have some special define instead of magic number. MPIDI_COLL_SELECT_AUTO or something like that.

File: src/mpid/ch4/src/ch4_coll_globals_default.c.in [hunk 7]
```diff
@@ -51,6 +57,13 @@ const MPIDI_coll_algo_container_t CH4_Bcast_intra_composition_gamma_cnt = {
                                                 }
 };
 
+const MPIDI_coll_algo_container_t CH4_Bcast_intra_composition_delta_cnt = {
+    .id = MPIDI_CH4_Bcast_intra_composition_delta_id,
+    .params.ch4_bcast_params.ch4_bcast_delta = {
+                                                .bcast = -1,
+                                                }
```
Comment (ddurnov) [inline src/mpid/ch4/src/ch4_coll_globals_default.c.in]: Hmmm. Looks like if you had one extra line, then you would make code_cleanup more adequate here. I would remove next like here or added next line in all other places :)

File: src/mpid/ch4/src/ch4_coll_impl.h [hunk 8]
```diff
@@ -581,6 +660,36 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_Reduce_intra_composition_beta(const void *sen
     goto fn_exit;
 }
 
+#ifdef MPIDI_BUILD_CH4_SHM
+#undef FUNCNAME
+#define FUNCNAME MPIDI_Reduce_intra_composition_gamma
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_Reduce_intra_composition_gamma(const void *sendbuf,
```
Comment (ddurnov) [inline src/mpid/ch4/src/ch4_coll_impl.h]: Compositions set is a pretty huge and relatively independent thing. I would move it into a separate commit

File: src/mpid/ch4/src/ch4_coll_select.h [hunk 9]
```diff
@@ -27,6 +27,7 @@ MPL_STATIC_INLINE_PREFIX
     if (comm->comm_kind == MPIR_COMM_KIND__INTERCOMM) {
         return (MPIDI_coll_algo_container_t *) & CH4_Barrier_inter_composition_alpha_cnt;
     }
+
```
Comment (ddurnov) [inline src/mpid/ch4/src/ch4_coll_select.h]: ?

File: src/include/mpir_coll.h [hunk 10]
```diff
@@ -390,22 +395,25 @@ int MPIR_Gatherv_impl(const void *sendbuf, int sendcnt, MPI_Datatype sendtype, v
                       MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag);
 
 /* intracomm-only functions */
-int MPIR_Gatherv_intra_auto(const void *sendbuf, int sendcnt, MPI_Datatype sendtype, void *recvbuf,
-                            const int *recvcnts, const int *displs, MPI_Datatype recvtype, int root,
-                            MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag);
+int MPIR_Gatherv_intra_auto(const void *sendbuf, int sendcnt, MPI_Datatype sendtype,
+                            void *recvbuf, const int *recvcnts, const int *displs,
+                            MPI_Datatype recvtype, int root, MPIR_Comm * comm_ptr,
+                            MPIR_Errflag_t * errflag);
```
[the remark this answers is not in the document]
Comment (ddurnov) [reply src/include/mpir_coll.h]: Yes, seems like it was caused by ```__``` + code_cleanup and ```__``` -> ```_``` + code_cleanup combination

File: src/mpid/ch4/netmod/ofi/ofi_coll.h (hunk 2 above)
Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_coll.h]: Yeah, I understand what you're doing here. I just don't know about adding the placeholder. What do you think @pavanbalaji ?

Comment (pavanbalaji) [reply src/mpid/ch4/netmod/ofi/ofi_coll.h]: I agree with @wesbland.  Adding the placeholder function is weird.

**Comment by AlexeyMalkhanov** [comment]
@pavanbalaji @raffenet @wesbland @ddurnov  I've created a separate PR https://github.com/pmodels/mpich/pull/2982 with keyword `allcomm` for communicator-type independent algorithms. This PR is self-consistent. Could you please take a look?

**Comment by AlexeyMalkhanov** [comment]
@pavanbalaji @raffenet @wesbland @ddurnov I've split the initial PR into small pieces. Please consider that commit 4e1312563a2ed9f3bbc04bc7b5eec9851cd70f4f ("Spreading collective selection infrastructure to all blocking collectives") is the latter in my local hierarchy but after the push it became second to the last.

**Comment by pavanbalaji** [comment]
@AlexeyMalkhanov the commit short message should be within 50 characters.

**Comment by AlexeyMalkhanov** [comment]
@pavanbalaji I've changed the commit short messages in accordance with the requirements. The commits are in correct order now. However, if it is more convenient I'll split this PR into several PRs which correspond to the commits.

**Comment by pavanbalaji** [comment]
@AlexeyMalkhanov I tried reviewing the patches and it looks like it merges from master, rather than rebases your patches based on master.  The current mpich workflow relies on rebasing, rather than merging.  Can you please rebase your commits on top of the latest master?

**Comment by AlexeyMalkhanov** [comment]
@pavanbalaji we follow the current mpich workflow and use rebase. We see some problems with commits ordering similar to described here:  https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/ . To align the order of commits we additionally made rebase on all our commits. Probably that reflected in the state you have observed. We did one more rebase from master and commits again seem to be in wrong order.

**Comment by pavanbalaji** [comment]
@AlexeyMalkhanov I look at the commits on my laptop, so I'm not too concerned about the order on github.  But without rebase, it's hard to figure out which commits I'm reviewing.

**Comment by AlexeyMalkhanov** [comment]
@pavanbalaji We've just rebased. Does it look fine now?

**Comment by pavanbalaji** [comment]
Are you sure.  I still see merged changes:

```
*   ac9af6d8ea25 (HEAD -> pr/collective_selection_infrastructure, AlexeyMalkhanov/collective_selection_infrastructure) Merge branch 'master' of https://github.com/pmodels/mpich into collective_selection_infra_all
|\  
| * 90a1690827ab romio: Enable system extentions
| * f3a49cd4604f romio: Include system headers later
| * ef684f80e505 mpl: Enable system extentions
| * 57a116ee8e1d mpl: Include system headers after mpl.h
* | eece07c14de1 CH4: added inter/intra prefixes to composition names
* | e3e8224e44b8 Collective names and indents cleanup
* | 62825084541c Collective selection infrastructure code_cleanup
* |   e3862002ba01 Merge branch 'master' of https://github.com/pmodels/mpich into collective_selection_infra_all
|\ \  
| |/  
| * 8ec7e370e27d ch3/nemesis: Remove unused code
| * 188c64019e50 mpl/shm: Add missing memory class for allocation in sysv impl
| * 11f037736601 mpl/shm: Fix function namespace in sysv impl
| * ea8f8b15326c romio: Initialize error_code before calling MPIO_DATATYPE_ISCOMMITTED
| * ebd09b845f16 romio: Fix shadow variable warning
| * 32fc80abe2bd ch4/ucx: Fix snprintf usage
| * aaf72961ef63 ch4/ofi: Honor max_msg_size when using iov
| * 47e86d9eab56 ch4/ofi: Refactor iov send/recv code
| * 0bd8f6cb2783 ch4/ofi: Move sync_send_ack bit out of protocol mask
* |   0ad268a7609c Collective selection infrastructure aligment with refactored collectives
```

**Comment by wesbland** [comment]
I'm not sure what you're seeing @pavanbalaji:

```
* 133b50238856 2018-01-22 Alexey Malkhanov (HEAD -> AlexeyMalkhanov-collective_selection_refactoring) coll: Selection infratructure for blocking collectives
* 34fb90bb3895 2018-02-15 Alexey Malkhanov coll: Unify names of structure fields for params
* dd2e43c9140c 2018-02-14 Alexey Malkhanov coll: Unify names for parameter containers
* 76d3ca6c06a6 2018-02-14 Alexey Malkhanov coll: Remove NM/SHM wrappers over MPIR algorithms
* 89385464faa3 2018-02-14 Alexey Malkhanov coll: Introduce compositions for shm algorithms
* cf68399a2ba4 2018-02-14 Alexey Malkhanov coll: Unify communicator variable name
* 85ba75c4c60e 2018-02-14 Alexey Malkhanov coll: Init variables in coll selection functions
* 3b072a9a3e2f 2018-02-14 Alexey Malkhanov coll: Remove separate call functions after select
* 61be43c948c6 2018-02-13 Alexey Malkhanov coll: Barrier recursive_doubling -> dissemination
* b53b6566212e 2018-02-13 Alexey Malkhanov coll: Unify names for containers and ids
* 98630b11f595 2018-02-13 Alexey Malkhanov coll: Fix for multinetmod build
* 28e97d6cd1f9 2018-02-13 Alexey Malkhanov coll: Rename file coll_tuning_types.h
* 918e50e2de95 2018-02-13 Alexey Malkhanov coll: Unify naming scheme for algorithms
* 52b0d18d3fce 2017-10-26 Surabhi Jain (argonne/master) dloop: Handled divide by zero cases
```

If you used `hub` to check out the original PR, sometimes it doesn't do a good job of updating the branch and you have to delete it and check it out again.

**Comment by pavanbalaji** [comment]
Thanks, @wesbland.  Removing the branch and recreating it fixed the problem.

**Comment by AlexeyMalkhanov** [comment]
@pavanbalaji I've updated the PR in accordance with your feedback: commits have been rearranged, namespace prefixes have been added.

**Comment by pavanbalaji** [comment]
I've signed off on all except two commits.  @raffenet can you review the remaining two?  They look fine to me, but I'd like a second opinion.

**Comment by raffenet** [comment]
Yes, I am looking at it.

File: src/mpid/ch4/src/ch4_coll.h [hunk 11]
```diff
@@ -34,6 +34,12 @@ MPL_STATIC_INLINE_PREFIX int MPID_Barrier(MPIR_Comm * comm, MPIR_Errflag_t * err
             mpi_errno =
                 MPIDI_Barrier_intra_composition_beta(comm, errflag, ch4_algo_parameters_container);
             break;
+#ifdef MPIDI_BUILD_CH4_SHM
+        case MPIDI_CH4_Barrier_intra_composition_gamma_id:
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_coll.h]: @AlexeyMalkhanov I'm having a little trouble picturing how this will work. Will the selection functions also need modification to actually use these compositions?

Comment (AlexeyMalkhanov) [reply src/mpid/ch4/src/ch4_coll.h]: @raffenet at the moment selection functions repeat legacy logic of collective selection. For the most collectives, there was no possibility to choose shm-specific algorithms. These shm-specific pieces in call phases can be treated as a way to handle shm algorithms which will be chosen based on decision tree parsing.

Comment (AlexeyMalkhanov) [reply src/mpid/ch4/src/ch4_coll.h]: @raffenet if you remember we previously introduced `is_single_node` attribute of a communicator for that purpose and based on the value of the attribute we chose netmod or shm inside a single composition which operates with the whole communicator. It turned out that the approach with `is_single_node` was dubious and we decided to introduce a separate composition for the case when the whole communicator resides inside one node.

Comment (raffenet) [reply src/mpid/ch4/src/ch4_coll.h]: I see. I think I understand the concept, but without part 2 it's hard to completely visualize. Would it be much trouble to drop this patch until there is code ready to take advantage of it?

File: src/mpid/ch4/netmod/ofi/ofi_coll_select.h [hunk 12]
```diff
@@ -94,4 +94,385 @@ MPL_STATIC_INLINE_PREFIX
...
+                                                              int recvcount,
+                                                              MPI_Datatype recvtype,
+                                                              int root,
+                                                              MPIR_Comm * comm,
+                                                              MPIR_Errflag_t * errflag,
+                                                              MPIDI_OFI_coll_algo_container_t *
+                                                              ch4_algo_parameters_container_in
+                                                              ATTRIBUTE((unused)))
+{
+    return (MPIDI_OFI_coll_algo_container_t *) & MPIDI_OFI_Gather_intra_binomial_cnt;
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_coll_select.h]: Why are we casting the return value here? This is already the type we are dealing with.

File: src/mpid/ch4/shm/glue/shm_coll.h [hunk 13]
```diff
@@ -65,15 +65,17 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_SHM_mpi_allreduce(const void *sendbuf, void *
...
+                                                     void *algo_parameters_container)
 {
     int ret;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_POSIX_MPI_ALLGATHER);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_POSIX_MPI_ALLGATHER);
 
     ret = MPIDI_POSIX_mpi_allgather(sendbuf, sendcount, sendtype, recvbuf, recvcount,
-                                    recvtype, comm, errflag);
+                                    recvtype, comm, errflag, (MPIDI_POSIX_coll_algo_container_t *)
```
Comment (raffenet) [inline src/mpid/ch4/shm/glue/shm_coll.h]: IMO, it would be much cleaner if the cast of this container took place in the POSIX coll select function. It's not interpreted between here and there, so I don't see why we should work to change the type so early. Am I missing something?

**Comment by raffenet** [comment]
@AlexeyMalkhanov what do you think about moving the remaining 2 commits into the phase 2 portion of this work? That way we can merge the non-controversial stuff from this PR and continue to iterate on the new stuff.

**Comment by pavanbalaji** [comment]
I agree with @raffenet.

@AlexeyMalkhanov @wesbland @ddurnov what do you guys think?  We'd like to wrap up the `3.3b2` pieces soon.  We have a bunch of other pending things remaining for that release.

File: src/mpid/ch4/netmod/ofi/ofi_coll_select.h (hunk 12 above)
Comment (AlexeyMalkhanov) [reply src/mpid/ch4/netmod/ofi/ofi_coll_select.h]: @raffenet ```MPIDI_OFI_Gather_intra_binomial_cnt``` is of type ```const MPIDI_OFI_coll_algo_container_t```, but function returns ```MPIDI_OFI_coll_algo_container_t *```. If we do not cast the return value here we get a warning. We could change the function prototypes, but it will be a large piece of differences to review. If the removing of the casting is critical we could introduce the appropriate infrastructure for the part 2 of collective selection.

File: src/mpid/ch4/shm/glue/shm_coll.h (hunk 13 above)
Comment (AlexeyMalkhanov) [reply src/mpid/ch4/shm/glue/shm_coll.h]: @raffenet Agree, will correct that

**Comment by AlexeyMalkhanov** [comment]
@pavanbalaji @raffenet @wesbland @ddurnov I have updated the PR in accordance with your feedback: 
1) the introduction of compositions has been removed from this PR and will appear in the second part of the functionality. These changes impacted on one commit that had been signed off already: https://github.com/pmodels/mpich/pull/2935/commits/ec6b11a8c8ee8421bdfdf318cf6c8421f1723b8b
2) Container types casting has been moved only to selection phase
IMO, the essence of this PR consists in spreading collective selection skeleton to all blocking collectives from the previously implemented structure for Barrier, Bcast, Reduce and Allreduce. All commits before the last one in the frames of this PR are just preliminary work for the infrastructure extension.
The second part of the collective selection infrastructure is supposed to influence mostly the selection part where we introduce new logic based on a decision tree.

**Comment by raffenet** [comment]
@AlexeyMalkhanov thanks, this makes it easier to understand. I'm going to test one more time as I review, but I don't see any major issues now. I'm glad we could remove those casts, too :smile:.
test:jenkins/ch4/most

**Comment by raffenet** [comment]
Oh the whitespace checker is complaining. I can help with that. I'm also going to split https://github.com/pmodels/mpich/commit/ec6b11a8c8ee8421bdfdf318cf6c8421f1723b8b into 2 commits to separate the container re-ordering from the addition of the constant.

**Comment by raffenet** [comment]
Added whitespace fixes and split one commit into two.
test:jenkins/ch4/most

**Comment by raffenet** [comment]
test:jenkins/ch4/most

**Comment by raffenet** [comment]
OK I think I got all the whitespace stuff now.
test:jenkins/ch4/most

**Comment by raffenet** [comment]
Couple of fixes, including a big rebase error I made in previous version.
test:jenkins/ch4/most

**Comment by raffenet** [comment]
@pavanbalaji @AlexeyMalkhanov I have one last minor nitpick. Can we change to commit header prefix to be `ch4/coll:`? This way it is clear this code does not affect upper layers or ch3. I will make the change if you agree.

**Comment by AlexeyMalkhanov** [comment]
@raffenet there is one commit https://github.com/pmodels/mpich/pull/2935/commits/44fe3dab23a101ef0ab2d2bde157f7efb71efcc4 which affects upper layers. As for the rest of commits, I completely agree with your proposal to change the header prefix.

**Comment by raffenet** [comment]
OK, I will take that into account. Will push/merge a new version shortly.
